



<!DOCTYPE html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      <meta http-equiv="x-ua-compatible" content="ie=edge">
      
        <meta name="description" content="Compressed storage for SNP data">
      
      
      
        <meta name="author" content="Hua Zhou">
      
      
        <meta name="lang:clipboard.copy" content="Copy to clipboard">
      
        <meta name="lang:clipboard.copied" content="Copied to clipboard">
      
        <meta name="lang:search.language" content="en">
      
        <meta name="lang:search.pipeline.stopwords" content="True">
      
        <meta name="lang:search.pipeline.trimmer" content="True">
      
        <meta name="lang:search.result.none" content="No matching documents">
      
        <meta name="lang:search.result.one" content="1 matching document">
      
        <meta name="lang:search.result.other" content="# matching documents">
      
        <meta name="lang:search.tokenizer" content="[\s\-]+">
      
      <link rel="shortcut icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-0.17.5, mkdocs-material-2.9.2">
    
    
      
        <title>SnpArray - SnpArrays.jl</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/application.ba0fd1a6.css">
      
      
    
    
      <script src="../../assets/javascripts/modernizr.1aa3b519.js"></script>
    
    
      <link href="https://fonts.gstatic.com" rel="preconnect" crossorigin>
      
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700|Roboto+Mono">
        <style>body,input{font-family:"Roboto","Helvetica Neue",Helvetica,Arial,sans-serif}code,kbd,pre{font-family:"Roboto Mono","Courier New",Courier,monospace}</style>
      
    
    <link rel="stylesheet" href="../../assets/fonts/material-icons.css">
    
      <link rel="stylesheet" href="../../assets/Documenter.css">
    
    
  </head>
  
    <body dir="ltr">
  
    <svg class="md-svg">
      <defs>
        
        
          <svg xmlns="http://www.w3.org/2000/svg" width="416" height="448"
    viewBox="0 0 416 448" id="__github">
  <path fill="currentColor" d="M160 304q0 10-3.125 20.5t-10.75 19-18.125
        8.5-18.125-8.5-10.75-19-3.125-20.5 3.125-20.5 10.75-19 18.125-8.5
        18.125 8.5 10.75 19 3.125 20.5zM320 304q0 10-3.125 20.5t-10.75
        19-18.125 8.5-18.125-8.5-10.75-19-3.125-20.5 3.125-20.5 10.75-19
        18.125-8.5 18.125 8.5 10.75 19 3.125 20.5zM360
        304q0-30-17.25-51t-46.75-21q-10.25 0-48.75 5.25-17.75 2.75-39.25
        2.75t-39.25-2.75q-38-5.25-48.75-5.25-29.5 0-46.75 21t-17.25 51q0 22 8
        38.375t20.25 25.75 30.5 15 35 7.375 37.25 1.75h42q20.5 0
        37.25-1.75t35-7.375 30.5-15 20.25-25.75 8-38.375zM416 260q0 51.75-15.25
        82.75-9.5 19.25-26.375 33.25t-35.25 21.5-42.5 11.875-42.875 5.5-41.75
        1.125q-19.5 0-35.5-0.75t-36.875-3.125-38.125-7.5-34.25-12.875-30.25-20.25-21.5-28.75q-15.5-30.75-15.5-82.75
        0-59.25 34-99-6.75-20.5-6.75-42.5 0-29 12.75-54.5 27 0 47.5 9.875t47.25
        30.875q36.75-8.75 77.25-8.75 37 0 70 8 26.25-20.5
        46.75-30.25t47.25-9.75q12.75 25.5 12.75 54.5 0 21.75-6.75 42 34 40 34
        99.5z" />
</svg>
        
      </defs>
    </svg>
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" data-md-component="overlay" for="__drawer"></label>
    
      <a href="#snparray" tabindex="1" class="md-skip">
        Skip to content
      </a>
    
    
      <header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid">
    <div class="md-flex">
      <div class="md-flex__cell md-flex__cell--shrink">
        <a href="../.." title="SnpArrays.jl" class="md-header-nav__button md-logo">
          
            <i class="md-icon"></i>
          
        </a>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        <label class="md-icon md-icon--menu md-header-nav__button" for="__drawer"></label>
      </div>
      <div class="md-flex__cell md-flex__cell--stretch">
        <div class="md-flex__ellipsis md-header-nav__title" data-md-component="title">
          
            
              <span class="md-header-nav__topic">
                SnpArrays.jl
              </span>
              <span class="md-header-nav__topic">
                SnpArray
              </span>
            
          
        </div>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        
          
            <label class="md-icon md-icon--search md-header-nav__button" for="__search"></label>
            
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="query" data-md-state="active">
      <label class="md-icon md-search__icon" for="__search"></label>
      <button type="reset" class="md-icon md-search__icon" data-md-component="reset" tabindex="-1">
        &#xE5CD;
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="result">
          <div class="md-search-result__meta">
            Type to start searching
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
          
        
      </div>
      
        <div class="md-flex__cell md-flex__cell--shrink">
          <div class="md-header-nav__source">
            


  


  <a href="https://github.com/OpenMendel/SnpArrays.jl/" title="Go to repository" class="md-source" data-md-source="github">
    
      <div class="md-source__icon">
        <svg viewBox="0 0 24 24" width="24" height="24">
          <use xlink:href="#__github" width="24" height="24"></use>
        </svg>
      </div>
    
    <div class="md-source__repository">
      GitHub
    </div>
  </a>

          </div>
        </div>
      
    </div>
  </nav>
</header>
    
    <div class="md-container">
      
        
      
      
      <main class="md-main">
        <div class="md-main__inner md-grid" data-md-component="container">
          
            
              <div class="md-sidebar md-sidebar--primary" data-md-component="navigation">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    <nav class="md-nav md-nav--primary" data-md-level="0">
  <label class="md-nav__title md-nav__title--site" for="__drawer">
    <span class="md-nav__button md-logo">
      
        <i class="md-icon"></i>
      
    </span>
    SnpArrays.jl
  </label>
  
    <div class="md-nav__source">
      


  


  <a href="https://github.com/OpenMendel/SnpArrays.jl/" title="Go to repository" class="md-source" data-md-source="github">
    
      <div class="md-source__icon">
        <svg viewBox="0 0 24 24" width="24" height="24">
          <use xlink:href="#__github" width="24" height="24"></use>
        </svg>
      </div>
    
    <div class="md-source__repository">
      GitHub
    </div>
  </a>

    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      


  <li class="md-nav__item">
    <a href="../.." title="Home" class="md-nav__link">
      Home
    </a>
  </li>

    
      
      
      

  


  <li class="md-nav__item md-nav__item--active md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-2" type="checkbox" id="nav-2" checked>
    
    <label class="md-nav__link" for="nav-2">
      Manual
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-2">
        Manual
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          

  


  <li class="md-nav__item md-nav__item--active">
    
    <input class="md-toggle md-nav__toggle" data-md-toggle="toc" type="checkbox" id="__toc">
    
      
    
    
      <label class="md-nav__link md-nav__link--active" for="__toc">
        SnpArray
      </label>
    
    <a href="./" title="SnpArray" class="md-nav__link md-nav__link--active">
      SnpArray
    </a>
    
      
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#constructor" title="Constructor" class="md-nav__link">
    Constructor
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#summary-statistics" title="Summary statistics" class="md-nav__link">
    Summary statistics
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#filtering" title="Filtering" class="md-nav__link">
    Filtering
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#random-genotypes-generation" title="Random genotypes generation" class="md-nav__link">
    Random genotypes generation
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#subsetting" title="Subsetting" class="md-nav__link">
    Subsetting
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#assignment" title="Assignment" class="md-nav__link">
    Assignment
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#convert-copy-and-imputation" title="Convert, copy and imputation" class="md-nav__link">
    Convert, copy and imputation
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#empirical-kinship-matrix" title="Empirical kinship matrix" class="md-nav__link">
    Empirical kinship matrix
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#principal-components" title="Principal components" class="md-nav__link">
    Principal components
  </a>
  
</li>
      
      
      
      
      
    </ul>
  
</nav>
    
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../snpdata/" title="SnpData" class="md-nav__link">
      SnpData
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../haplotypearray/" title="HaplotypeArray" class="md-nav__link">
      HaplotypeArray
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              <div class="md-sidebar md-sidebar--secondary" data-md-component="toc">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#constructor" title="Constructor" class="md-nav__link">
    Constructor
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#summary-statistics" title="Summary statistics" class="md-nav__link">
    Summary statistics
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#filtering" title="Filtering" class="md-nav__link">
    Filtering
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#random-genotypes-generation" title="Random genotypes generation" class="md-nav__link">
    Random genotypes generation
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#subsetting" title="Subsetting" class="md-nav__link">
    Subsetting
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#assignment" title="Assignment" class="md-nav__link">
    Assignment
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#convert-copy-and-imputation" title="Convert, copy and imputation" class="md-nav__link">
    Convert, copy and imputation
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#empirical-kinship-matrix" title="Empirical kinship matrix" class="md-nav__link">
    Empirical kinship matrix
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#principal-components" title="Principal components" class="md-nav__link">
    Principal components
  </a>
  
</li>
      
      
      
      
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/OpenMendel/SnpArrays.jl/edit/master/docs/man/snparray.md" title="Edit this page" class="md-icon md-content__icon">&#xE3C9;</a>
                
                
                <p><a id='SnpArray-1'></a></p>
<h1 id="snparray">SnpArray</h1>
<p><code>SnpArray</code> is an array of <code>Tuple{Bool,Bool}</code> and adopts the same coding as the <a href="http://zzz.bwh.harvard.edu/plink/binary.shtml">Plink binary format</a>. If <code>A1</code> and <code>A2</code> are the two alleles, the coding rule is  </p>
<table>
<thead>
<tr>
<th align="center">Genotype</th>
<th align="center">SnpArray</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center">A1,A1</td>
<td align="center">(false,false)</td>
</tr>
<tr>
<td align="center">A1,A2</td>
<td align="center">(false,true)</td>
</tr>
<tr>
<td align="center">A2,A2</td>
<td align="center">(true,true)</td>
</tr>
<tr>
<td align="center">missing</td>
<td align="center">(true,false)</td>
</tr>
</tbody>
</table>
<p>The code <code>(true,false)</code> is reserved for missing genotype. Otherwise, the bit <code>true</code> represents one copy of allele <code>A2</code>. In a two-dimensional <code>SnpArray</code>, each row is a person and each column is a SNP.</p>
<p>For complete genotype data, for example, after imputation, consider using the <a href="../haplotypearray/#HaplotypeArray-1">HaplotypeArray</a> type.</p>
<p><a id='Constructor-1'></a></p>
<h2 id="constructor">Constructor</h2>
<p>There are various ways to initialize a <code>SnpArray</code>.  </p>
<ul>
<li><code>SnpArray</code> can be initialized from <a href="http://zzz.bwh.harvard.edu/plink/binary.shtml">Plink binary files</a>, say the sample data set <code>hapmap3</code> in package <code>docs</code> folder:</li>
</ul>
<div class="codehilite"><pre><span></span><span class="p">;</span><span class="n">ls</span> <span class="o">-</span><span class="n">l</span> <span class="o">$</span><span class="p">(</span><span class="n">Pkg</span><span class="o">.</span><span class="n">dir</span><span class="p">(</span><span class="s">&quot;SnpArrays&quot;</span><span class="p">)</span> <span class="o">*</span> <span class="s">&quot;/docs/hapmap3.*&quot;</span><span class="p">)</span>
</pre></div>


<div class="codehilite"><pre><span></span>-rw-r--r--  1 huazhou  staff  1128171 Jun 19  2017 /Users/huazhou/.julia/v0.6/SnpArrays/docs/hapmap3.bed
-rw-r--r--  1 huazhou  staff   388672 Jun 19  2017 /Users/huazhou/.julia/v0.6/SnpArrays/docs/hapmap3.bim
-rw-r--r--  1 huazhou  staff     7136 Jun 19  2017 /Users/huazhou/.julia/v0.6/SnpArrays/docs/hapmap3.fam
-rw-r--r--  1 huazhou  staff   332960 Jun 19  2017 /Users/huazhou/.julia/v0.6/SnpArrays/docs/hapmap3.map
</pre></div>


<div class="codehilite"><pre><span></span><span class="k">using</span> <span class="n">SnpArrays</span>
<span class="n">hapmap</span> <span class="o">=</span> <span class="n">SnpArray</span><span class="p">(</span><span class="n">Pkg</span><span class="o">.</span><span class="n">dir</span><span class="p">(</span><span class="s">&quot;SnpArrays&quot;</span><span class="p">)</span> <span class="o">*</span> <span class="s">&quot;/docs/hapmap3&quot;</span><span class="p">)</span>
</pre></div>


<div class="codehilite"><pre><span></span>[1m[36mINFO: [39m[22m[36mv1.0 BED file detected
[39m




324×13928 SnpArrays.SnpArray{2}:
 (true, true)  (true, true)   (false, false)  …  (true, true)   (true, true)
 (true, true)  (false, true)  (false, true)      (false, true)  (true, true)
 (true, true)  (true, true)   (false, true)      (true, true)   (true, true)
 (true, true)  (true, true)   (false, true)      (true, true)   (true, true)
 (true, true)  (true, true)   (false, true)      (true, true)   (true, true)
 (true, true)  (false, true)  (true, true)    …  (true, true)   (true, true)
 (true, true)  (true, true)   (true, true)       (true, true)   (true, true)
 (true, true)  (true, true)   (false, false)     (true, true)   (true, true)
 (true, true)  (true, true)   (false, true)      (true, true)   (true, true)
 (true, true)  (true, true)   (false, true)      (true, true)   (true, true)
 (true, true)  (true, true)   (false, true)   …  (true, true)   (true, true)
 (true, true)  (true, true)   (true, true)       (true, true)   (true, true)
 (true, true)  (true, true)   (false, false)     (true, true)   (true, true)
 ⋮                                            ⋱                             
 (true, true)  (true, true)   (false, true)      (true, true)   (true, true)
 (true, true)  (true, true)   (false, false)     (true, true)   (true, true)
 (true, true)  (true, true)   (false, false)     (true, true)   (true, true)
 (true, true)  (true, true)   (false, false)  …  (true, true)   (true, true)
 (true, true)  (true, true)   (false, true)      (true, true)   (true, true)
 (true, true)  (true, true)   (true, true)       (true, true)   (true, true)
 (true, true)  (true, true)   (false, true)      (true, true)   (true, true)
 (true, true)  (true, true)   (false, true)      (true, true)   (true, true)
 (true, true)  (true, true)   (false, false)  …  (true, true)   (true, true)
 (true, true)  (true, true)   (false, true)      (true, true)   (true, true)
 (true, true)  (true, true)   (false, false)     (true, true)   (true, true)
 (true, true)  (true, true)   (false, false)     (true, true)   (true, true)
</pre></div>


<p>By default, the constructor figures out the number of individuals and SNPs from the <code>.bim</code> and <code>.fam</code> files.</p>
<div class="codehilite"><pre><span></span><span class="c"># rows are people; columns are SNPs</span>
<span class="n">people</span><span class="p">,</span> <span class="n">snps</span> <span class="o">=</span> <span class="n">size</span><span class="p">(</span><span class="n">hapmap</span><span class="p">)</span>
</pre></div>


<div class="codehilite"><pre><span></span>(324, 13928)
</pre></div>


<p>Alternatively, users can supply keyword arguments <code>people</code> and <code>snps</code> to the constructor. In this case only the <code>.bed</code> file needs to be present.</p>
<div class="codehilite"><pre><span></span><span class="n">hapmap</span> <span class="o">=</span> <span class="n">SnpArray</span><span class="p">(</span><span class="n">Pkg</span><span class="o">.</span><span class="n">dir</span><span class="p">(</span><span class="s">&quot;SnpArrays&quot;</span><span class="p">)</span> <span class="o">*</span> <span class="s">&quot;/docs/hapmap3&quot;</span><span class="p">;</span> <span class="n">people</span> <span class="o">=</span> <span class="mi">324</span><span class="p">,</span> <span class="n">snps</span> <span class="o">=</span> <span class="mi">13928</span><span class="p">)</span>
</pre></div>


<div class="codehilite"><pre><span></span>[1m[36mINFO: [39m[22m[36mv1.0 BED file detected
[39m




324×13928 SnpArrays.SnpArray{2}:
 (true, true)  (true, true)   (false, false)  …  (true, true)   (true, true)
 (true, true)  (false, true)  (false, true)      (false, true)  (true, true)
 (true, true)  (true, true)   (false, true)      (true, true)   (true, true)
 (true, true)  (true, true)   (false, true)      (true, true)   (true, true)
 (true, true)  (true, true)   (false, true)      (true, true)   (true, true)
 (true, true)  (false, true)  (true, true)    …  (true, true)   (true, true)
 (true, true)  (true, true)   (true, true)       (true, true)   (true, true)
 (true, true)  (true, true)   (false, false)     (true, true)   (true, true)
 (true, true)  (true, true)   (false, true)      (true, true)   (true, true)
 (true, true)  (true, true)   (false, true)      (true, true)   (true, true)
 (true, true)  (true, true)   (false, true)   …  (true, true)   (true, true)
 (true, true)  (true, true)   (true, true)       (true, true)   (true, true)
 (true, true)  (true, true)   (false, false)     (true, true)   (true, true)
 ⋮                                            ⋱                             
 (true, true)  (true, true)   (false, true)      (true, true)   (true, true)
 (true, true)  (true, true)   (false, false)     (true, true)   (true, true)
 (true, true)  (true, true)   (false, false)     (true, true)   (true, true)
 (true, true)  (true, true)   (false, false)  …  (true, true)   (true, true)
 (true, true)  (true, true)   (false, true)      (true, true)   (true, true)
 (true, true)  (true, true)   (true, true)       (true, true)   (true, true)
 (true, true)  (true, true)   (false, true)      (true, true)   (true, true)
 (true, true)  (true, true)   (false, true)      (true, true)   (true, true)
 (true, true)  (true, true)   (false, false)  …  (true, true)   (true, true)
 (true, true)  (true, true)   (false, true)      (true, true)   (true, true)
 (true, true)  (true, true)   (false, false)     (true, true)   (true, true)
 (true, true)  (true, true)   (false, false)     (true, true)   (true, true)
</pre></div>


<p>Internally <code>SnpArray</code> stores data as <code>BitArray</code>s and consumes approximately the same amount of memory as the Plink <code>bed</code> file size.</p>
<div class="codehilite"><pre><span></span><span class="c"># memory usage, bed file size</span>
<span class="n">Base</span><span class="o">.</span><span class="n">summarysize</span><span class="p">(</span><span class="n">hapmap</span><span class="p">),</span> <span class="n">filesize</span><span class="p">(</span><span class="s">&quot;hapmap3.bed&quot;</span><span class="p">)</span>
</pre></div>


<div class="codehilite"><pre><span></span>(1128256, 1128171)
</pre></div>


<ul>
<li><code>SnpArray</code> can be initialized from a matrix of A1 allele counts.</li>
</ul>
<div class="codehilite"><pre><span></span><span class="n">SnpArray</span><span class="p">(</span><span class="n">rand</span><span class="p">(</span><span class="mi">0</span><span class="o">:</span><span class="mi">2</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>
</pre></div>


<div class="codehilite"><pre><span></span>5×3 SnpArrays.SnpArray{2}:
 (true, true)    (true, true)   (false, false)
 (false, false)  (true, true)   (false, true) 
 (false, false)  (false, true)  (false, true) 
 (false, false)  (false, true)  (true, true)  
 (true, true)    (true, true)   (true, true)
</pre></div>


<ul>
<li><code>SnpArray(m, n)</code> generates an m by n <code>SnpArray</code> of all A1 alleles.</li>
</ul>
<div class="codehilite"><pre><span></span><span class="n">s</span> <span class="o">=</span> <span class="n">SnpArray</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>
</pre></div>


<div class="codehilite"><pre><span></span>5×3 SnpArrays.SnpArray{2}:
 (false, false)  (false, false)  (false, false)
 (false, false)  (false, false)  (false, false)
 (false, false)  (false, false)  (false, false)
 (false, false)  (false, false)  (false, false)
 (false, false)  (false, false)  (false, false)
</pre></div>


<p><a id='Summary-statistics-1'></a></p>
<h2 id="summary-statistics">Summary statistics</h2>
<p><code>summarize</code> function computes the following summary statistics of a <code>SnpArray</code>:  </p>
<ul>
<li><code>maf</code>: minor allele frequencies, taking into account of missingness.</li>
<li><code>minor_allele</code>: a <code>BitVector</code> indicating the minor allele for each SNP.   <code>minor_allele[j]==true</code> means A1 is the minor allele for SNP j; <code>minor_allele[j]==false</code> means A2 is the minor allele for SNP j.</li>
<li><code>missings_by_snp</code>: number of missing genotypes for each snp.</li>
<li><code>missings_by_person</code>: number of missing genotypes for each person.</li>
</ul>
<div class="codehilite"><pre><span></span><span class="n">maf</span><span class="p">,</span> <span class="n">minor_allele</span><span class="p">,</span> <span class="n">missings_by_snp</span><span class="p">,</span> <span class="n">missings_by_person</span> <span class="o">=</span> <span class="n">summarize</span><span class="p">(</span><span class="n">hapmap</span><span class="p">)</span>
<span class="c"># minor allele frequencies</span>
<span class="n">maf</span><span class="o">&#39;</span>
</pre></div>


<div class="codehilite"><pre><span></span>1×13928 RowVector{Float64,Array{Float64,1}}:
 0.0  0.0776398  0.324074  0.191589  …  0.00154321  0.0417957  0.00617284
</pre></div>


<div class="codehilite"><pre><span></span><span class="c"># total number of missing genotypes</span>
<span class="n">sum</span><span class="p">(</span><span class="n">missings_by_snp</span><span class="p">),</span> <span class="n">sum</span><span class="p">(</span><span class="n">missings_by_person</span><span class="p">)</span>
</pre></div>


<div class="codehilite"><pre><span></span>(11894, 11894)
</pre></div>


<div class="codehilite"><pre><span></span><span class="c"># proportion of missing genotypes</span>
<span class="n">sum</span><span class="p">(</span><span class="n">missings_by_snp</span><span class="p">)</span> <span class="o">/</span> <span class="n">length</span><span class="p">(</span><span class="n">hapmap</span><span class="p">)</span>
</pre></div>


<div class="codehilite"><pre><span></span>0.0026356890108565393
</pre></div>


<p><a id='Filtering-1'></a></p>
<h2 id="filtering">Filtering</h2>
<p>In almost all analyses, SNPs and individuals with low genotyping success rates are ignored. This filtering step is an important tool for removing likely false positives from association testing, as genotyping failure often occurs preferentially in cases or controls, or is correlated with the quantitative trait. <code>filter(s, min_success_rate_per_snp, min_success_rate_per_person)</code> does filtering according to the specified success rates for SNPs and people. Default is 0.98 for both.</p>
<div class="codehilite"><pre><span></span><span class="c"># filtering SNPs and people to have both success rates above 0.98</span>
<span class="n">snp_idx</span><span class="p">,</span> <span class="n">person_idx</span> <span class="o">=</span> <span class="n">filter</span><span class="p">(</span><span class="n">hapmap</span><span class="p">,</span> <span class="mf">0.98</span><span class="p">,</span> <span class="mf">0.98</span><span class="p">)</span>
<span class="c"># summary statistics of the filtered SnpArray</span>
<span class="n">_</span><span class="p">,</span> <span class="n">_</span><span class="p">,</span> <span class="n">missings_by_snp_filtered</span><span class="p">,</span> <span class="n">missings_by_person_filtered</span> <span class="o">=</span> <span class="n">summarize</span><span class="p">(</span><span class="n">hapmap</span><span class="p">[</span><span class="n">person_idx</span><span class="p">,</span> <span class="n">snp_idx</span><span class="p">]);</span>
</pre></div>


<div class="codehilite"><pre><span></span><span class="c"># minimum SNP genotyping success rate after filtering ≥ 0.98</span>
<span class="mf">1.0</span> <span class="o">-</span> <span class="n">maximum</span><span class="p">(</span><span class="n">missings_by_snp_filtered</span><span class="p">)</span> <span class="o">/</span> <span class="n">length</span><span class="p">(</span><span class="n">missings_by_person_filtered</span><span class="p">)</span>
</pre></div>


<div class="codehilite"><pre><span></span>0.9813084112149533
</pre></div>


<div class="codehilite"><pre><span></span><span class="c"># minimum person genotyping success rate after filtering ≥ 0.98</span>
<span class="mf">1.0</span> <span class="o">-</span> <span class="n">maximum</span><span class="p">(</span><span class="n">missings_by_person_filtered</span><span class="p">)</span> <span class="o">/</span> <span class="n">length</span><span class="p">(</span><span class="n">missings_by_snp_filtered</span><span class="p">)</span>
</pre></div>


<div class="codehilite"><pre><span></span>0.9818511796733213
</pre></div>


<p><a id='Random-genotypes-generation-1'></a></p>
<h2 id="random-genotypes-generation">Random genotypes generation</h2>
<p><code>randgeno(a1freq)</code> generates a random genotype according to A1 allele frequency <code>a1freq</code>.</p>
<div class="codehilite"><pre><span></span><span class="n">randgeno</span><span class="p">(</span><span class="mf">0.5</span><span class="p">)</span>
</pre></div>


<div class="codehilite"><pre><span></span>(true, true)
</pre></div>


<p><code>randgeno(maf, minor_allele)</code> generates a random genotype according to minor allele frequency <code>maf</code> and whether the minor allele is A1 (<code>minor_allele==true</code>) or A2 (<code>minor_allele==false</code>).</p>
<div class="codehilite"><pre><span></span><span class="n">randgeno</span><span class="p">(</span><span class="mf">0.25</span><span class="p">,</span> <span class="kc">true</span><span class="p">)</span>
</pre></div>


<div class="codehilite"><pre><span></span>(false, true)
</pre></div>


<p><code>randgeno(n, maf, minor_allele)</code> generates a vector of random genotypes according to a common minor allele frequency <code>maf</code> and the minor allele.</p>
<div class="codehilite"><pre><span></span><span class="n">randgeno</span><span class="p">(</span><span class="mi">10</span><span class="p">,</span> <span class="mf">0.25</span><span class="p">,</span> <span class="kc">true</span><span class="p">)</span>
</pre></div>


<div class="codehilite"><pre><span></span>10-element SnpArrays.SnpArray{1}:
 (false, true) 
 (false, true) 
 (true, true)  
 (true, true)  
 (true, true)  
 (true, true)  
 (true, true)  
 (true, true)  
 (false, false)
 (true, true)
</pre></div>


<p><code>randgeno(m, n, maf, minor_allele)</code> generates a random $m$-by-$n$ <code>SnpArray</code> according to a vector of minor allele frequencies <code>maf</code> and a minor allele indicator vector. The lengths of both vectors should be <code>n</code>.</p>
<div class="codehilite"><pre><span></span><span class="c"># this is a random replicate of the hapmap data</span>
<span class="n">randgeno</span><span class="p">(</span><span class="n">size</span><span class="p">(</span><span class="n">hapmap</span><span class="p">),</span> <span class="n">maf</span><span class="p">,</span> <span class="n">minor_allele</span><span class="p">)</span>
</pre></div>


<div class="codehilite"><pre><span></span>324×13928 SnpArrays.SnpArray{2}:
 (true, true)  (true, true)    …  (true, true)   (true, true) 
 (true, true)  (true, true)       (true, true)   (true, true) 
 (true, true)  (false, true)      (false, true)  (true, true) 
 (true, true)  (false, false)     (true, true)   (true, true) 
 (true, true)  (false, true)      (true, true)   (true, true) 
 (true, true)  (true, true)    …  (true, true)   (true, true) 
 (true, true)  (true, true)       (true, true)   (true, true) 
 (true, true)  (true, true)       (true, true)   (true, true) 
 (true, true)  (true, true)       (true, true)   (true, true) 
 (true, true)  (true, true)       (true, true)   (true, true) 
 (true, true)  (false, true)   …  (true, true)   (true, true) 
 (true, true)  (true, true)       (true, true)   (true, true) 
 (true, true)  (true, true)       (true, true)   (true, true) 
 ⋮                             ⋱                              
 (true, true)  (true, true)       (true, true)   (true, true) 
 (true, true)  (true, true)       (true, true)   (false, true)
 (true, true)  (true, true)       (true, true)   (true, true) 
 (true, true)  (false, true)   …  (true, true)   (true, true) 
 (true, true)  (true, true)       (false, true)  (true, true) 
 (true, true)  (false, true)      (true, true)   (true, true) 
 (true, true)  (true, true)       (true, true)   (true, true) 
 (true, true)  (true, true)       (true, true)   (true, true) 
 (true, true)  (false, true)   …  (true, true)   (true, true) 
 (true, true)  (true, true)       (true, true)   (true, true) 
 (true, true)  (true, true)       (false, true)  (true, true) 
 (true, true)  (true, true)       (true, true)   (true, true)
</pre></div>


<p><a id='Subsetting-1'></a></p>
<h2 id="subsetting">Subsetting</h2>
<p>Subsetting a <code>SnpArray</code> works the same way as subsetting any other arrays.</p>
<div class="codehilite"><pre><span></span><span class="c"># genotypes of the 1st person</span>
<span class="n">hapmap</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="o">:</span><span class="p">]</span>
</pre></div>


<div class="codehilite"><pre><span></span>13928-element SnpArrays.SnpArray{1}:
 (true, true)  
 (true, true)  
 (false, false)
 (true, true)  
 (true, true)  
 (true, true)  
 (false, true) 
 (false, true) 
 (true, true)  
 (false, true) 
 (true, true)  
 (true, true)  
 (false, false)
 ⋮             
 (false, true) 
 (false, true) 
 (true, true)  
 (false, true) 
 (false, true) 
 (false, true) 
 (false, true) 
 (false, true) 
 (false, true) 
 (true, true)  
 (true, true)  
 (true, true)
</pre></div>


<div class="codehilite"><pre><span></span><span class="c"># genotypes of the 5th SNP</span>
<span class="n">hapmap</span><span class="p">[</span><span class="o">:</span><span class="p">,</span> <span class="mi">5</span><span class="p">]</span>
</pre></div>


<div class="codehilite"><pre><span></span>324-element SnpArrays.SnpArray{1}:
 (true, true)  
 (true, true)  
 (false, true) 
 (false, true) 
 (true, true)  
 (false, false)
 (false, false)
 (true, true)  
 (true, true)  
 (true, true)  
 (true, true)  
 (true, true)  
 (false, true) 
 ⋮             
 (false, false)
 (true, true)  
 (false, true) 
 (true, true)  
 (true, true)  
 (true, true)  
 (true, true)  
 (true, true)  
 (false, true) 
 (true, true)  
 (true, true)  
 (true, true)
</pre></div>


<div class="codehilite"><pre><span></span><span class="c"># subsetting both persons and SNPs</span>
<span class="n">hapmap</span><span class="p">[</span><span class="mi">1</span><span class="o">:</span><span class="mi">5</span><span class="p">,</span> <span class="mi">5</span><span class="o">:</span><span class="mi">10</span><span class="p">]</span>
</pre></div>


<div class="codehilite"><pre><span></span>5×6 SnpArrays.SnpArray{2}:
 (true, true)   (true, true)  (false, true)  …  (true, true)   (false, true)
 (true, true)   (true, true)  (true, true)      (true, true)   (false, true)
 (false, true)  (true, true)  (true, true)      (false, true)  (true, true) 
 (false, true)  (true, true)  (true, true)      (true, true)   (false, true)
 (true, true)   (true, true)  (true, true)      (true, true)   (false, true)
</pre></div>


<div class="codehilite"><pre><span></span><span class="c"># filter out rare SNPs with MAF &lt; 0.05</span>
<span class="n">hapmap</span><span class="p">[</span><span class="o">:</span><span class="p">,</span> <span class="n">maf</span> <span class="o">.≥</span> <span class="mf">0.05</span><span class="p">]</span>
</pre></div>


<div class="codehilite"><pre><span></span>324×12085 SnpArrays.SnpArray{2}:
 (true, true)   (false, false)  …  (false, true)  (false, true)
 (false, true)  (false, true)      (true, true)   (true, true) 
 (true, true)   (false, true)      (true, true)   (true, true) 
 (true, true)   (false, true)      (false, true)  (false, true)
 (true, true)   (false, true)      (true, true)   (true, true) 
 (false, true)  (true, true)    …  (false, true)  (false, true)
 (true, true)   (true, true)       (true, true)   (true, true) 
 (true, true)   (false, false)     (true, true)   (true, true) 
 (true, true)   (false, true)      (true, true)   (true, true) 
 (true, true)   (false, true)      (false, true)  (false, true)
 (true, true)   (false, true)   …  (true, true)   (true, true) 
 (true, true)   (true, true)       (false, true)  (false, true)
 (true, true)   (false, false)     (false, true)  (false, true)
 ⋮                              ⋱                              
 (true, true)   (false, true)      (false, true)  (false, true)
 (true, true)   (false, false)     (false, true)  (false, true)
 (true, true)   (false, false)     (true, true)   (true, true) 
 (true, true)   (false, false)  …  (true, true)   (true, true) 
 (true, true)   (false, true)      (true, true)   (true, true) 
 (true, true)   (true, true)       (false, true)  (false, true)
 (true, true)   (false, true)      (false, true)  (false, true)
 (true, true)   (false, true)      (true, true)   (true, true) 
 (true, true)   (false, false)  …  (false, true)  (false, true)
 (true, true)   (false, true)      (false, true)  (false, true)
 (true, true)   (false, false)     (false, true)  (false, true)
 (true, true)   (false, false)     (true, true)   (true, true)
</pre></div>


<div class="codehilite"><pre><span></span><span class="c"># filter out individuals with genotyping success rate &lt; 0.90</span>
<span class="n">hapmap</span><span class="p">[</span><span class="n">missings_by_person</span> <span class="o">/</span> <span class="n">people</span> <span class="o">.&lt;</span> <span class="mf">0.1</span><span class="p">,</span> <span class="o">:</span><span class="p">]</span>
</pre></div>


<div class="codehilite"><pre><span></span>220×13928 SnpArrays.SnpArray{2}:
 (true, true)  (true, true)   (false, false)  …  (true, true)   (true, true)
 (true, true)  (false, true)  (false, true)      (false, true)  (true, true)
 (true, true)  (true, true)   (false, true)      (true, true)   (true, true)
 (true, true)  (true, true)   (true, true)       (true, true)   (true, true)
 (true, true)  (true, true)   (false, false)     (true, true)   (true, true)
 (true, true)  (true, true)   (false, true)   …  (true, true)   (true, true)
 (true, true)  (true, true)   (false, true)      (true, true)   (true, true)
 (true, true)  (true, true)   (false, true)      (true, true)   (true, true)
 (true, true)  (true, true)   (false, false)     (true, true)   (true, true)
 (true, true)  (true, true)   (false, true)      (true, true)   (true, true)
 (true, true)  (false, true)  (false, true)   …  (true, true)   (true, true)
 (true, true)  (true, true)   (true, true)       (true, true)   (true, true)
 (true, true)  (true, true)   (true, true)       (true, true)   (true, true)
 ⋮                                            ⋱                             
 (true, true)  (true, true)   (false, false)     (true, true)   (true, true)
 (true, true)  (true, true)   (false, false)     (true, true)   (true, true)
 (true, true)  (true, true)   (false, false)  …  (true, true)   (true, true)
 (true, true)  (true, true)   (false, true)      (true, true)   (true, true)
 (true, true)  (true, true)   (false, false)     (true, true)   (true, true)
 (true, true)  (true, true)   (false, false)     (true, true)   (true, true)
 (true, true)  (true, true)   (false, false)     (true, true)   (true, true)
 (true, true)  (true, true)   (false, true)   …  (true, true)   (true, true)
 (true, true)  (true, true)   (false, false)     (true, true)   (true, true)
 (true, true)  (true, true)   (false, true)      (true, true)   (true, true)
 (true, true)  (true, true)   (false, false)     (true, true)   (true, true)
 (true, true)  (true, true)   (false, false)     (true, true)   (true, true)
</pre></div>


<p><code>view</code> creates views of subarray without copying data and improve efficiency in many calculations.</p>
<div class="codehilite"><pre><span></span><span class="n">mafcommon</span><span class="p">,</span> <span class="o">=</span> <span class="n">summarize</span><span class="p">(</span><span class="n">view</span><span class="p">(</span><span class="n">hapmap</span><span class="p">,</span> <span class="o">:</span><span class="p">,</span> <span class="n">maf</span> <span class="o">.≥</span> <span class="mf">0.05</span><span class="p">))</span>
<span class="n">mafcommon</span><span class="o">&#39;</span>
</pre></div>


<div class="codehilite"><pre><span></span>1×12085 RowVector{Float64,Array{Float64,1}}:
 0.0776398  0.324074  0.191589  …  0.310937  0.23913  0.23913  0.23913
</pre></div>


<p><a id='Assignment-1'></a></p>
<h2 id="assignment">Assignment</h2>
<p>It is possible to assign specific genotypes to a <code>SnpArray</code> entry.</p>
<div class="codehilite"><pre><span></span><span class="n">hapmap</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span>
</pre></div>


<div class="codehilite"><pre><span></span>(true, true)
</pre></div>


<div class="codehilite"><pre><span></span><span class="n">hapmap</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span><span class="kc">false</span><span class="p">,</span> <span class="kc">true</span><span class="p">)</span>
<span class="n">hapmap</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span>
</pre></div>


<div class="codehilite"><pre><span></span>(false, true)
</pre></div>


<div class="codehilite"><pre><span></span><span class="n">hapmap</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="nb">NaN</span>
<span class="n">hapmap</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span>
</pre></div>


<div class="codehilite"><pre><span></span>(true, false)
</pre></div>


<div class="codehilite"><pre><span></span><span class="n">hapmap</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="mi">2</span>
<span class="n">hapmap</span><span class="p">[</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span>
</pre></div>


<div class="codehilite"><pre><span></span>(true, true)
</pre></div>


<p>Subsetted assignment such as <code>hapmap[:, 1] = Nan</code> is also valid.</p>
<p><a id='Convert,-copy-and-imputation-1'></a></p>
<h2 id="convert-copy-and-imputation">Convert, copy and imputation</h2>
<p>In most analyses we convert a whole <code>SnpArray</code> or slices of it to numeric arrays (matrix of <strong>minor allele counts</strong>) for statistical analysis. Keep in mind that the storage of resultant data can be up to 32 fold larger than that of the original <code>SnpArray</code>. Fortunately, rich collection of data types in <code>Julia</code> allow us choose one that fits into memory. Below are estimates of memory usage for some common data types with <code>n</code> persons and <code>p</code> SNPs. Here MAF denotes the <strong>average</strong> minor allele frequencies.</p>
<ul>
<li><code>SnpArray</code>: $0.25np$ bytes</li>
<li><code>Matrix{Int8}</code>: $np$ bytes</li>
<li><code>Matrix{Float16}</code>: $2np$ bytes</li>
<li><code>Matrix{Float32}</code>: $4np$ bytes</li>
<li><code>Matrix{Float64}</code>: $8np$ bytes</li>
<li><code>SparseMatrixCSC{Float64,Int64}</code>: $16 \cdot \text{NNZ} + 8(p+1) \approx 16np(2\text{MAF}(1-\text{MAF})+\text{MAF}^2) + 8(p+1) = 16np \cdot \text{MAF}(2-\text{MAF}) + 8(p+1)$ bytes. When the average MAF=0.25, this is about $7np$ bytes. When MAF=0.025, this is about $0.8np$ bypes, 10 fold smaller than the <code>Matrix{Float64}</code> type.</li>
<li><code>SparseMatrixCSC{Int8,UInt32}</code>: $5 \cdot \text{NNZ} + 4(p+1) \approx 5np(2\text{MAF}(1-\text{MAF})+\text{MAF}^2) + 4(p+1) = 5np \cdot \text{MAF}(2-\text{MAF}) + 4(p+1)$ bytes. When the average MAF=0.25, this is about $2.2np$ bytes. When MAF=0.08, this is about $0.8np$ bypes, 10 fold smaller than <code>Matrix{Float64}</code> type.</li>
<li>Two <code>SparseMatrixCSC{Bool,Int64}</code>: $2np \cdot \text{MAF} \cdot 9 + 16(p+1) = 18 np \cdot \text{MAF} + 16(p+1)$ bytes. When the average MAF=0.25, this is about $4.5np$ bytes. When MAF=0.045, this is about $0.8np$ bytes, 10 fold smaller than <code>Matrix{Float64}</code> type.</li>
</ul>
<p>To be concrete, consider 2 typical data sets:  </p>
<ul>
<li>COPD (GWAS): $n = 6670$ individuals, $p = 630998$ SNPs, average MAF is 0.2454.</li>
<li>GAW19 (sequencing study): $n = 959$ individuals, $p = 8348674$ SNPs, average MAF is 0.085.</li>
</ul>
<table>
<thead>
<tr>
<th align="right">Data Type</th>
<th align="right">COPD</th>
<th align="right">GAW19</th>
</tr>
</thead>
<tbody>
<tr>
<td align="right"><code>SnpArray</code></td>
<td align="right">1.05GB</td>
<td align="right">2GB</td>
</tr>
<tr>
<td align="right"><code>Matrix{Float64}</code></td>
<td align="right">33.67GB</td>
<td align="right">64.05GB</td>
</tr>
<tr>
<td align="right"><code>SparseMatrixCSC{Float64,Int64}</code></td>
<td align="right">29GB</td>
<td align="right">20.82GB</td>
</tr>
<tr>
<td align="right"><code>SparseMatrixCSC{Bool,Int64}</code></td>
<td align="right">18.6GB</td>
<td align="right">12.386GB</td>
</tr>
</tbody>
</table>
<p>Apparently for data sets with a majority of rare variants, converting to sparse matrices saves memory and often brings computational advantages too. In the <code>SparseMatrixCSC</code> format, the integer type of the row indices <code>rowval</code> and column pointer <code>colptr</code> should have maximal allowable value larger than the number of nonzeros in the matrix. The <code>InexactError()</code> error encountered during conversion often indicates that the integer type has a too small range. The utility function <code>estimatesize</code> conveniently estimates memory usage in bytes for the input data type.</p>
<div class="codehilite"><pre><span></span><span class="c"># estimated memory usage if convert to Matrix{Float64}</span>
<span class="n">estimatesize</span><span class="p">(</span><span class="n">people</span><span class="p">,</span> <span class="n">snps</span><span class="p">,</span> <span class="kt">Matrix</span><span class="p">{</span><span class="kt">Float64</span><span class="p">})</span>
</pre></div>


<div class="codehilite"><pre><span></span>3.6101376e7
</pre></div>


<div class="codehilite"><pre><span></span><span class="c"># convert to Matrix{Float64}</span>
<span class="n">hapmapf64</span> <span class="o">=</span> <span class="n">convert</span><span class="p">(</span><span class="kt">Matrix</span><span class="p">{</span><span class="kt">Float64</span><span class="p">},</span> <span class="n">hapmap</span><span class="p">)</span>
</pre></div>


<div class="codehilite"><pre><span></span>324×13928 Array{Float64,2}:
 0.0  0.0  2.0  0.0  0.0  0.0  1.0  1.0  …  1.0  1.0  1.0  1.0  0.0  0.0  0.0
 0.0  1.0  1.0  1.0  0.0  0.0  0.0  1.0     0.0  0.0  0.0  0.0  0.0  1.0  0.0
 0.0  0.0  1.0  1.0  1.0  0.0  0.0  2.0     1.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  1.0  0.0  1.0  0.0  0.0  1.0     1.0  1.0  1.0  1.0  0.0  0.0  0.0
 0.0  0.0  1.0  1.0  0.0  0.0  0.0  2.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  1.0  0.0  0.0  2.0  0.0  0.0  0.0  …  2.0  1.0  1.0  1.0  0.0  0.0  0.0
 0.0  0.0  0.0  0.0  2.0  0.0  0.0  2.0     1.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  2.0  0.0  0.0  0.0  0.0  1.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  1.0  1.0  0.0  0.0  0.0  2.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  1.0  0.0  0.0  0.0  0.0  2.0     1.0  1.0  1.0  1.0  0.0  0.0  0.0
 0.0  0.0  1.0  0.0  0.0  0.0  0.0  2.0  …  0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  1.0  0.0  0.0  0.0  2.0     1.0  1.0  1.0  1.0  0.0  0.0  0.0
 0.0  0.0  2.0  0.0  1.0  0.0  0.0  1.0     2.0  1.0  1.0  1.0  0.0  0.0  0.0
 ⋮                        ⋮              ⋱                      ⋮            
 0.0  0.0  1.0  0.0  2.0  0.0  0.0  1.0     2.0  1.0  1.0  1.0  0.0  0.0  0.0
 0.0  0.0  2.0  0.0  0.0  0.0  0.0  2.0     2.0  1.0  1.0  1.0  0.0  0.0  0.0
 0.0  0.0  2.0  0.0  1.0  0.0  0.0  1.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  2.0  1.0  0.0  0.0  0.0  1.0  …  0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  1.0  0.0  0.0  0.0  0.0  1.0     1.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0     2.0  1.0  1.0  1.0  0.0  0.0  0.0
 0.0  0.0  1.0  1.0  0.0  0.0  0.0  2.0     1.0  1.0  1.0  1.0  0.0  0.0  0.0
 0.0  0.0  1.0  0.0  0.0  0.0  0.0  2.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0
 0.0  0.0  2.0  0.0  1.0  0.0  0.0  1.0  …  1.0  1.0  1.0  1.0  0.0  0.0  0.0
 0.0  0.0  1.0  1.0  0.0  0.0  0.0  1.0     1.0  1.0  1.0  1.0  0.0  0.0  0.0
 0.0  0.0  2.0  1.0  0.0  0.0  0.0  2.0     1.0  1.0  1.0  1.0  0.0  0.0  0.0
 0.0  0.0  2.0  0.0  0.0  0.0  0.0  2.0     0.0  0.0  0.0  0.0  0.0  0.0  0.0
</pre></div>


<div class="codehilite"><pre><span></span><span class="c"># actual memory usage of Matrix{Float64}</span>
<span class="n">Base</span><span class="o">.</span><span class="n">summarysize</span><span class="p">(</span><span class="n">hapmapf64</span><span class="p">)</span>
</pre></div>


<div class="codehilite"><pre><span></span>36101376
</pre></div>


<div class="codehilite"><pre><span></span><span class="c"># average maf of the hapmap3 data set</span>
<span class="n">mean</span><span class="p">(</span><span class="n">maf</span><span class="p">)</span>
</pre></div>


<div class="codehilite"><pre><span></span>0.222585591341583
</pre></div>


<div class="codehilite"><pre><span></span><span class="c"># estimated memory usage if convert to SparseMatrixCSC{Float32, UInt32} matrix</span>
<span class="n">estimatesize</span><span class="p">(</span><span class="n">people</span><span class="p">,</span> <span class="n">snps</span><span class="p">,</span> <span class="kt">SparseMatrixCSC</span><span class="p">{</span><span class="kt">Float32</span><span class="p">,</span> <span class="kt">UInt32</span><span class="p">},</span> <span class="n">mean</span><span class="p">(</span><span class="n">maf</span><span class="p">))</span>
</pre></div>


<div class="codehilite"><pre><span></span>1.4338389205819245e7
</pre></div>


<div class="codehilite"><pre><span></span><span class="c"># convert to SparseMatrixCSC{Float32, UInt32} matrix</span>
<span class="n">hapmapf32sp</span> <span class="o">=</span> <span class="n">convert</span><span class="p">(</span><span class="kt">SparseMatrixCSC</span><span class="p">{</span><span class="kt">Float32</span><span class="p">,</span> <span class="kt">UInt32</span><span class="p">},</span> <span class="n">hapmap</span><span class="p">)</span>
</pre></div>


<div class="codehilite"><pre><span></span>324×13928 SparseMatrixCSC{Float32,UInt32} with 1614876 stored entries:
  [2    ,     2]  =  1.0
  [6    ,     2]  =  1.0
  [15   ,     2]  =  1.0
  [31   ,     2]  =  1.0
  [33   ,     2]  =  1.0
  [35   ,     2]  =  1.0
  [43   ,     2]  =  1.0
  [44   ,     2]  =  1.0
  [50   ,     2]  =  1.0
  [54   ,     2]  =  1.0
  ⋮
  [135  , 13927]  =  1.0
  [148  , 13927]  =  1.0
  [160  , 13927]  =  1.0
  [164  , 13927]  =  2.0
  [167  , 13927]  =  1.0
  [185  , 13927]  =  1.0
  [266  , 13927]  =  1.0
  [280  , 13927]  =  1.0
  [288  , 13927]  =  1.0
  [118  , 13928]  =  2.0
  [231  , 13928]  =  2.0
</pre></div>


<div class="codehilite"><pre><span></span><span class="c"># actual memory usage if convert to SparseMatrixCSC{Float32, UInt32} matrix</span>
<span class="n">Base</span><span class="o">.</span><span class="n">summarysize</span><span class="p">(</span><span class="n">hapmapf32sp</span><span class="p">)</span>
</pre></div>


<div class="codehilite"><pre><span></span>12974764
</pre></div>


<p>By default the <code>convert()</code> method converts missing genotypes to <code>NaN</code>.</p>
<div class="codehilite"><pre><span></span><span class="c"># number of missing genotypes</span>
<span class="n">countnz</span><span class="p">(</span><span class="n">isnan</span><span class="o">.</span><span class="p">(</span><span class="n">hapmap</span><span class="p">)),</span> <span class="n">countnz</span><span class="p">(</span><span class="n">isnan</span><span class="o">.</span><span class="p">(</span><span class="n">hapmapf64</span><span class="p">))</span>
</pre></div>


<div class="codehilite"><pre><span></span>(11894, 11894)
</pre></div>


<p>One can enforce <strong>crude imputation</strong> by setting the optional argument <code>impute=true</code>. Imputation is done by generating two random alleles according to the minor allele frequency. This is a neutral but not an optimal strategy, and users should impute missing genotypes by more advanced methods.</p>
<div class="codehilite"><pre><span></span><span class="n">hapmapf64impute</span> <span class="o">=</span> <span class="n">convert</span><span class="p">(</span><span class="kt">Matrix</span><span class="p">{</span><span class="kt">Float64</span><span class="p">},</span> <span class="n">hapmap</span><span class="p">;</span> <span class="n">impute</span> <span class="o">=</span> <span class="kc">true</span><span class="p">)</span>
<span class="n">countnz</span><span class="p">(</span><span class="n">isnan</span><span class="o">.</span><span class="p">(</span><span class="n">hapmapf64impute</span><span class="p">))</span>
</pre></div>


<div class="codehilite"><pre><span></span>0
</pre></div>


<p>By default <code>convert()</code> translates genotypes according to the <em>additive</em> SNP model, which essentially counts the number of <strong>minor allele</strong> (0, 1 or 2) per genotype. Other SNP models are <em>dominant</em> and <em>recessive</em>, both in terms of the <strong>minor allele</strong>. When <code>A1</code> is the minor allele, genotypes are translated to real number according to</p>
<table>
<thead>
<tr>
<th align="center">Genotype</th>
<th align="center"><code>SnpArray</code></th>
<th align="center"><code>model=:additive</code></th>
<th align="center"><code>model=:dominant</code></th>
<th align="center"><code>model=:recessive</code></th>
</tr>
</thead>
<tbody>
<tr>
<td align="center">A1,A1</td>
<td align="center">(false,false)</td>
<td align="center">2</td>
<td align="center">1</td>
<td align="center">1</td>
</tr>
<tr>
<td align="center">A1,A2</td>
<td align="center">(false,true)</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">0</td>
</tr>
<tr>
<td align="center">A2,A2</td>
<td align="center">(true,true)</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
<tr>
<td align="center">missing</td>
<td align="center">(true,false)</td>
<td align="center">NaN</td>
<td align="center">NaN</td>
<td align="center">NaN</td>
</tr>
</tbody>
</table>
<p>When <code>A2</code> is the minor allele, genotypes are translated according to</p>
<table>
<thead>
<tr>
<th align="center">Genotype</th>
<th align="center"><code>SnpArray</code></th>
<th align="center"><code>model=:additive</code></th>
<th align="center"><code>model=:dominant</code></th>
<th align="center"><code>model=:recessive</code></th>
</tr>
</thead>
<tbody>
<tr>
<td align="center">A1,A1</td>
<td align="center">(false,false)</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
<tr>
<td align="center">A1,A2</td>
<td align="center">(false,true)</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">0</td>
</tr>
<tr>
<td align="center">A2,A2</td>
<td align="center">(true,true)</td>
<td align="center">2</td>
<td align="center">1</td>
<td align="center">1</td>
</tr>
<tr>
<td align="center">missing</td>
<td align="center">(true,false)</td>
<td align="center">NaN</td>
<td align="center">NaN</td>
<td align="center">NaN</td>
</tr>
</tbody>
</table>
<div class="codehilite"><pre><span></span><span class="p">[</span><span class="n">convert</span><span class="p">(</span><span class="kt">Vector</span><span class="p">{</span><span class="kt">Float64</span><span class="p">},</span> <span class="n">hapmap</span><span class="p">[</span><span class="mi">1</span><span class="o">:</span><span class="mi">10</span><span class="p">,</span> <span class="mi">5</span><span class="p">];</span> <span class="n">model</span> <span class="o">=</span> <span class="o">:</span><span class="n">additive</span><span class="p">)</span> <span class="n">convert</span><span class="p">(</span><span class="kt">Vector</span><span class="p">{</span><span class="kt">Float64</span><span class="p">},</span> <span class="n">hapmap</span><span class="p">[</span><span class="mi">1</span><span class="o">:</span><span class="mi">10</span><span class="p">,</span> <span class="mi">5</span><span class="p">];</span> <span class="n">model</span> <span class="o">=</span> <span class="o">:</span><span class="n">dominant</span><span class="p">)</span> <span class="n">convert</span><span class="p">(</span><span class="kt">Vector</span><span class="p">{</span><span class="kt">Float64</span><span class="p">},</span> <span class="n">hapmap</span><span class="p">[</span><span class="mi">1</span><span class="o">:</span><span class="mi">10</span><span class="p">,</span> <span class="mi">5</span><span class="p">];</span> <span class="n">model</span> <span class="o">=</span> <span class="o">:</span><span class="n">recessive</span><span class="p">)]</span>
</pre></div>


<div class="codehilite"><pre><span></span>10×3 Array{Float64,2}:
 0.0  0.0  0.0
 0.0  0.0  0.0
 1.0  1.0  0.0
 1.0  1.0  0.0
 0.0  0.0  0.0
 2.0  1.0  1.0
 2.0  1.0  1.0
 0.0  0.0  0.0
 0.0  0.0  0.0
 0.0  0.0  0.0
</pre></div>


<p>By default <code>convert()</code> does <strong>not</strong> center and scale genotypes. Setting the optional arguments <code>center=true, scale=true</code> centers genotypes at 2MAF and scales them by $[2 \cdot \text{MAF} \cdot (1 - \text{MAF})]^{-1/2}$. Mono-allelic SNPs (MAF=0) are not scaled.</p>
<div class="codehilite"><pre><span></span><span class="p">[</span><span class="n">convert</span><span class="p">(</span><span class="kt">Vector</span><span class="p">{</span><span class="kt">Float64</span><span class="p">},</span> <span class="n">hapmap</span><span class="p">[</span><span class="o">:</span><span class="p">,</span> <span class="mi">5</span><span class="p">])</span> <span class="n">convert</span><span class="p">(</span><span class="kt">Vector</span><span class="p">{</span><span class="kt">Float64</span><span class="p">},</span> <span class="n">hapmap</span><span class="p">[</span><span class="o">:</span><span class="p">,</span> <span class="mi">5</span><span class="p">];</span> <span class="n">center</span> <span class="o">=</span> <span class="kc">true</span><span class="p">,</span> <span class="n">scale</span> <span class="o">=</span> <span class="kc">true</span><span class="p">)]</span>
</pre></div>


<div class="codehilite"><pre><span></span>324×2 Array{Float64,2}:
 0.0  -1.25702 
 0.0  -1.25702 
 1.0   0.167017
 1.0   0.167017
 0.0  -1.25702 
 2.0   1.59106 
 2.0   1.59106 
 0.0  -1.25702 
 0.0  -1.25702 
 0.0  -1.25702 
 0.0  -1.25702 
 0.0  -1.25702 
 1.0   0.167017
 ⋮             
 2.0   1.59106 
 0.0  -1.25702 
 1.0   0.167017
 0.0  -1.25702 
 0.0  -1.25702 
 0.0  -1.25702 
 0.0  -1.25702 
 0.0  -1.25702 
 1.0   0.167017
 0.0  -1.25702 
 0.0  -1.25702 
 0.0  -1.25702
</pre></div>


<p><code>copy!()</code> is the in-place version of <code>convert()</code>. Options such as GWAS loop over SNPs and perform statistical anlaysis for each SNP. This can be achieved by</p>
<div class="codehilite"><pre><span></span><span class="n">g</span> <span class="o">=</span> <span class="n">zeros</span><span class="p">(</span><span class="n">people</span><span class="p">)</span>
<span class="k">for</span> <span class="n">j</span> <span class="o">=</span> <span class="mi">1</span><span class="o">:</span><span class="n">snps</span>
    <span class="n">copy!</span><span class="p">(</span><span class="n">g</span><span class="p">,</span> <span class="n">hapmap</span><span class="p">[</span><span class="o">:</span><span class="p">,</span> <span class="n">j</span><span class="p">];</span> <span class="n">model</span> <span class="o">=</span> <span class="o">:</span><span class="n">additive</span><span class="p">,</span> <span class="n">impute</span> <span class="o">=</span> <span class="kc">true</span><span class="p">)</span>
    <span class="c"># do statistical anlaysis</span>
<span class="k">end</span>
</pre></div>


<p><a id='Empirical-kinship-matrix-1'></a></p>
<h2 id="empirical-kinship-matrix">Empirical kinship matrix</h2>
<p><code>grm</code> function computes the empirical kinship matrix using either the genetic relationship matrix, <code>grm(A, model=:GRM)</code>, or the method of moment method, <code>grm(A, model=:MoM)</code>. </p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Missing genotypes are imputed according to minor allele frequencies on the fly.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>It is often necessary to filter SNPs according to minor allele frequency and LD before calculating empirical kinship matrix.</p>
</div>
<p>By default, <code>grm</code> exlcude SNPs with minor allele frequency below 0.01. This can be changed by the keyword argument <code>maf_threshold</code>.</p>
<div class="codehilite"><pre><span></span><span class="c"># GRM using all SNPs with MAF ≥ 0.01. Same as</span>
<span class="c"># grm(hapmap; maf_threshold = 0.01)</span>
<span class="n">grm</span><span class="p">(</span><span class="n">hapmap</span><span class="p">)</span>
</pre></div>


<div class="codehilite"><pre><span></span>324×324 Array{Float64,2}:
 0.571688   0.0462807  0.0192437  …  0.0650573  0.0716031  0.0653311
 0.0462807  0.544895   0.0286567     0.051311   0.0449147  0.0629484
 0.0192437  0.0286567  0.520718      0.0460941  0.0297353  0.0362593
 0.0486305  0.0369559  0.0293866     0.0599803  0.0659248  0.0599897
 0.052624   0.0423282  0.0247818     0.0715649  0.0581852  0.0658928
 0.0441266  0.0319019  0.0378722  …  0.0697301  0.0563314  0.065507 
 0.0397138  0.0217914  0.0120808     0.0446991  0.0381401  0.0371412
 0.0410787  0.0382596  0.021239      0.0577011  0.0547265  0.065843 
 0.030168   0.0315418  0.0169194     0.0334515  0.0460561  0.0380265
 0.0394529  0.0422534  0.0258607     0.0665238  0.0574226  0.0496232
 0.0477793  0.0454803  0.0230092  …  0.0585435  0.0630797  0.0610807
 0.0607075  0.0387026  0.0370969     0.0669321  0.0590524  0.0709323
 0.0357632  0.0438016  0.0264945     0.0571821  0.0701542  0.0635772
 ⋮                                ⋱                                 
 0.0583693  0.0556652  0.0382274     0.0864722  0.0864521  0.0971764
 0.0676922  0.0521073  0.0324018     0.091956   0.0803552  0.0859155
 0.0679075  0.0558743  0.0386209     0.0899731  0.0844895  0.0857414
 0.0652022  0.0564352  0.0390764  …  0.0878313  0.0765041  0.0892029
 0.063271   0.0561774  0.0364754     0.0789625  0.0734351  0.0822378
 0.0648145  0.0612811  0.040045      0.0896104  0.0901562  0.0812354
 0.0625796  0.0600032  0.0380649     0.0831302  0.0852875  0.0809756
 0.0622714  0.0629473  0.0334698     0.0932245  0.0900163  0.0869392
 0.0660527  0.0633805  0.0322493  …  0.0973118  0.0791239  0.0866666
 0.0650573  0.051311   0.0460941     0.60326    0.0819594  0.0901512
 0.0716031  0.0449147  0.0297353     0.0819594  0.595108   0.0823624
 0.0653311  0.0629484  0.0362593     0.0901512  0.0823624  0.582986
</pre></div>


<div class="codehilite"><pre><span></span><span class="c"># GRM using all SNPs with MAF ≥ 0.05</span>
<span class="n">grm</span><span class="p">(</span><span class="n">hapmap</span><span class="p">;</span> <span class="n">maf_threshold</span> <span class="o">=</span> <span class="mf">0.05</span><span class="p">)</span>
</pre></div>


<div class="codehilite"><pre><span></span>324×324 Array{Float64,2}:
 0.569625   0.0535704  0.0232651  …  0.0737173  0.077806   0.0728392
 0.0535704  0.542874   0.0292732     0.0564599  0.0502244  0.0685582
 0.0232651  0.0292732  0.52172       0.0518046  0.0350828  0.0406411
 0.0530896  0.0398713  0.0315497     0.0621363  0.0726467  0.0656261
 0.0552446  0.0461061  0.0260241     0.0764474  0.0638611  0.0722666
 0.0491941  0.0332782  0.0418767  …  0.075888   0.0612478  0.0647859
 0.0426535  0.0262393  0.0149532     0.0467589  0.0381211  0.043397 
 0.044185   0.0420889  0.0242141     0.0573316  0.0569442  0.0666711
 0.0312286  0.0339761  0.0202836     0.0375401  0.0499576  0.041732 
 0.0418681  0.0447897  0.0290518     0.0711012  0.0638013  0.0531209
 0.0551049  0.047964   0.0237454  …  0.0628456  0.068017   0.0672129
 0.0663155  0.042911   0.0417686     0.0743179  0.0654391  0.0735138
 0.0359789  0.0473879  0.0296654     0.0629452  0.0732141  0.0676924
 ⋮                                ⋱                                 
 0.0609321  0.05769    0.0424677     0.0922023  0.0921072  0.105169 
 0.0720513  0.0550647  0.0373036     0.0979529  0.0860333  0.0926531
 0.0673758  0.0596123  0.0407195     0.0922654  0.090949   0.0909565
 0.0706774  0.064258   0.0456276  …  0.0946531  0.0831749  0.093832 
 0.0700268  0.0596294  0.0393517     0.0864101  0.0816786  0.0898796
 0.0692969  0.0634394  0.0440346     0.0938751  0.0951635  0.0888866
 0.0663192  0.0625808  0.0413768     0.0858972  0.0889115  0.0878044
 0.068348   0.067889   0.039336      0.100281   0.0920632  0.0865986
 0.0721291  0.0674194  0.0388886  …  0.106512   0.0845076  0.0901555
 0.0737173  0.0564599  0.0518046     0.59046    0.0903191  0.0945729
 0.077806   0.0502244  0.0350828     0.0903191  0.588007   0.0883015
 0.0728392  0.0685582  0.0406411     0.0945729  0.0883015  0.570052
</pre></div>


<div class="codehilite"><pre><span></span><span class="c"># GRM using every other SNP, with maf ≥ 0.01</span>
<span class="n">grm</span><span class="p">(</span><span class="n">view</span><span class="p">(</span><span class="n">hapmap</span><span class="p">,</span> <span class="o">:</span><span class="p">,</span> <span class="mi">1</span><span class="o">:</span><span class="mi">2</span><span class="o">:</span><span class="n">snps</span><span class="p">))</span>
</pre></div>


<div class="codehilite"><pre><span></span>324×324 Array{Float64,2}:
 0.55891    0.042845   0.027682   …  0.0677721  0.0746666  0.0687635
 0.042845   0.558745   0.0286764     0.0578987  0.0462962  0.0569583
 0.027682   0.0286764  0.512307      0.0390934  0.0396354  0.0476762
 0.045369   0.0469047  0.028284      0.0515663  0.0621996  0.0564942
 0.0523796  0.0477215  0.0267944     0.0676464  0.0566912  0.0656048
 0.0523564  0.03959    0.040644   …  0.0775905  0.0623572  0.0523832
 0.0393927  0.0262359  0.0178367     0.0471618  0.0347238  0.0343188
 0.047151   0.0379639  0.0263928     0.0590008  0.0559667  0.0688052
 0.0269253  0.0227214  0.0201293     0.0331204  0.0438332  0.0359987
 0.0323098  0.0402429  0.0238774     0.0598208  0.0408401  0.0505469
 0.0494403  0.0504749  0.0211352  …  0.0633784  0.0648717  0.0534096
 0.0626913  0.0494432  0.0407554     0.0668062  0.0648524  0.0606319
 0.0256349  0.0433059  0.0229496     0.0547733  0.0687123  0.063634 
 ⋮                                ⋱                                 
 0.0605549  0.0506841  0.0403012     0.0871397  0.0847116  0.101929 
 0.0663559  0.0599439  0.0310283     0.0894028  0.0762304  0.0813586
 0.0694759  0.0543091  0.0394317     0.0846661  0.0820957  0.096536 
 0.0654772  0.0581756  0.0404721  …  0.0934607  0.0768118  0.0828469
 0.0646551  0.0613316  0.0423482     0.0735479  0.0703157  0.0875644
 0.066537   0.0632911  0.0413394     0.0763028  0.082904   0.0844916
 0.0638155  0.0649174  0.032502      0.0869326  0.0919846  0.0835989
 0.0598241  0.0613027  0.0346562     0.0946688  0.0910173  0.0840148
 0.076463   0.0668378  0.0405983  …  0.101928   0.0747384  0.0896328
 0.0677721  0.0578987  0.0390934     0.573711   0.0700193  0.0959231
 0.0746666  0.0462962  0.0396354     0.0700193  0.57723    0.0752888
 0.0687635  0.0569583  0.0476762     0.0959231  0.0752888  0.588204
</pre></div>


<div class="codehilite"><pre><span></span><span class="c"># MoM using all SNPs with MAF ≥ 0.01</span>
<span class="n">grm</span><span class="p">(</span><span class="n">hapmap</span><span class="p">;</span> <span class="n">method</span> <span class="o">=</span> <span class="o">:</span><span class="n">MoM</span><span class="p">)</span>
</pre></div>


<div class="codehilite"><pre><span></span>324×324 Array{Float64,2}:
 0.539321    0.0346778  0.00253189  …  0.0537053  0.0637509  0.0503962
 0.0346778   0.517693   0.0141139      0.0421233  0.0395233  0.0499234
 0.00253189  0.0141139  0.499493       0.0320777  0.0210867  0.0188412
 0.0436597   0.0291232  0.0233322      0.0526417  0.0696601  0.0496871
 0.0456688   0.0333778  0.0156502      0.0658782  0.0570144  0.0564235
 0.0317232   0.0197867  0.0261686   …  0.0587872  0.0496871  0.0488598
 0.0243958   0.0104502  0.00276825     0.0313686  0.0305414  0.0238049
 0.0256958   0.0285322  0.00879561     0.0384596  0.0460234  0.0462598
 0.0220322   0.0260504  0.0135229      0.0298323  0.0488598  0.0301868
 0.0209685   0.0255777  0.0104502      0.0526417  0.0508689  0.0319596
 0.0362142   0.0330232  0.00525011  …  0.0496871  0.0561872  0.0500416
 0.0483871   0.0313686  0.0268777      0.0546508  0.0546508  0.051578 
 0.0313686   0.0415324  0.0238049      0.0585508  0.0704874  0.0577235
 ⋮                                  ⋱                                 
 0.0416506   0.041887   0.0254595      0.0713146  0.0792329  0.081833 
 0.0620963   0.0482689  0.0267595      0.0917604  0.0792329  0.0836057
 0.0552417   0.0453143  0.0256958      0.0806511  0.0817148  0.0727328
 0.0454325   0.0421233  0.0235685   …  0.072851   0.0628054  0.0703692
 0.0558326   0.0534689  0.0297141      0.0775784  0.070251   0.0736783
 0.0520507   0.0542962  0.0312505      0.0788784  0.0866785  0.0714328
 0.0522871   0.0507507  0.0295959      0.0772238  0.0827784  0.0732056
 0.0482689   0.0521689  0.0196685      0.0820693  0.0838421  0.0720237
 0.0500416   0.0477961  0.0214412   …  0.0880967  0.0735601  0.0693055
 0.0537053   0.0421233  0.0320777      0.561303   0.0807693  0.0755692
 0.0637509   0.0395233  0.0210867      0.0807693  0.566858   0.0688328
 0.0503962   0.0499234  0.0188412      0.0755692  0.0688328  0.533766
</pre></div>


<p><a id='Principal-components-1'></a></p>
<h2 id="principal-components">Principal components</h2>
<p>Principal compoenent analysis is widely used in genome-wide association analysis (GWAS) for adjusting population substructure. <code>pca(A, pcs)</code> computes the top <code>pcs</code> principal components of a <code>SnpArray</code>. Each SNP is centered at $2\text{MAF}$ and scaled by $[2\text{MAF}(1-\text{MAF})]^{-1/2}$. The output is  </p>
<ul>
<li><code>pcscore</code>: top <code>pcs</code> eigen-SNPs, or principal scores, in each column</li>
<li><code>pcloading</code>: top <code>pcs</code> eigen-vectors, or principal loadings, in each column</li>
<li><code>pcvariance</code>: top <code>pcs</code> eigen-values, or principal variances</li>
</ul>
<p>Missing genotypes are imputed according the minor allele frequencies on the fly. This implies that, in the presence of missing genotypes, running the function on the same <code>SnpArray</code> twice may produce slightly different answers. For reproducibility, it is a good practice to set the random seed before each function that does imputation on the fly.</p>
<div class="codehilite"><pre><span></span><span class="n">srand</span><span class="p">(</span><span class="mi">123</span><span class="p">)</span> <span class="c"># set seed</span>
<span class="n">pcscore</span><span class="p">,</span> <span class="n">pcloading</span><span class="p">,</span> <span class="n">pcvariance</span> <span class="o">=</span> <span class="n">pca</span><span class="p">(</span><span class="n">hapmap</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>
</pre></div>


<div class="codehilite"><pre><span></span>([-38.7231 -1.2983 -7.00541; -32.6096 -1.21052 -3.3232; … ; -48.9263 -2.06102 2.17374; -48.8627 0.274894 6.49518], [2.56616e-19 8.19569e-19 5.52006e-19; 0.00143962 -0.0042375 -0.00311816; … ; 0.00313326 -0.00427486 -0.0152038; -9.09523e-5 -0.00287777 0.0037855], [1841.4, 225.324, 70.7084])
</pre></div>


<p>To use eigen-SNPs for plotting or as covariates in GWAS, we typically scale them by their standard deviations so that they have mean zero and unit variance.</p>
<div class="codehilite"><pre><span></span><span class="c"># standardize eigen-SNPs before plotting or GWAS</span>
<span class="n">scale!</span><span class="p">(</span><span class="n">pcscore</span><span class="p">,</span> <span class="mi">1</span> <span class="o">./</span> <span class="n">sqrt</span><span class="o">.</span><span class="p">(</span><span class="n">pcvariance</span><span class="p">))</span>
<span class="n">std</span><span class="p">(</span><span class="n">pcscore</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
</pre></div>


<div class="codehilite"><pre><span></span>1×3 Array{Float64,2}:
 1.0  1.0  1.0
</pre></div>


<p>Internally <code>pca</code> converts <code>SnpArray</code> to the matrix of minor allele counts. The default format is <code>Matrix{Float64}</code>, which can easily exceed memory limit. Users have several options when the default <code>Matrix{Float64}</code> cannot fit into memory.  </p>
<ul>
<li>Use other intermediate matrix types.</li>
</ul>
<div class="codehilite"><pre><span></span><span class="c"># use single precision matrix and display the principal variances</span>
<span class="c"># approximately same answer as double precision</span>
<span class="n">srand</span><span class="p">(</span><span class="mi">123</span><span class="p">)</span>
<span class="n">pca</span><span class="p">(</span><span class="n">hapmap</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="kt">Matrix</span><span class="p">{</span><span class="kt">Float32</span><span class="p">})[</span><span class="mi">3</span><span class="p">]</span>
</pre></div>


<div class="codehilite"><pre><span></span>3-element Array{Float32,1}:
 1841.39  
  225.324 
   70.7084
</pre></div>


<ul>
<li>Use subset of SNPs</li>
</ul>
<div class="codehilite"><pre><span></span><span class="c"># principal components using every other SNP capture about half the variance</span>
<span class="n">srand</span><span class="p">(</span><span class="mi">123</span><span class="p">)</span>
<span class="n">pca</span><span class="p">(</span><span class="n">view</span><span class="p">(</span><span class="n">hapmap</span><span class="p">,</span> <span class="o">:</span><span class="p">,</span> <span class="mi">1</span><span class="o">:</span><span class="mi">2</span><span class="o">:</span><span class="n">snps</span><span class="p">),</span> <span class="mi">3</span><span class="p">)[</span><span class="mi">3</span><span class="p">]</span>
</pre></div>


<div class="codehilite"><pre><span></span>3-element Array{Float64,1}:
 926.622 
 113.188 
  36.4866
</pre></div>


<ul>
<li>Use sparse matrix. For large data sets with majority of rare variants, <code>pca_sp</code> is more efficient by first converting <code>SnpArray</code> to a sparse matrix (default is <code>SparseMatrixCSC{Float64, Int64}</code>) and then computing principal components using iterative algorithms.</li>
</ul>
<div class="codehilite"><pre><span></span><span class="c"># approximately same answer if we use Float16 sparse matrix</span>
<span class="n">srand</span><span class="p">(</span><span class="mi">123</span><span class="p">)</span>
<span class="n">pca_sp</span><span class="p">(</span><span class="n">hapmap</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="kt">SparseMatrixCSC</span><span class="p">{</span><span class="kt">Float32</span><span class="p">,</span> <span class="kt">UInt32</span><span class="p">})[</span><span class="mi">3</span><span class="p">]</span>
</pre></div>


<div class="codehilite"><pre><span></span>3-element Array{Float32,1}:
 1841.39  
  225.324 
   70.7084
</pre></div>
                
                  
                
              
              
                


              
            </article>
          </div>
        </div>
      </main>
      
        
<footer class="md-footer">
  
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid">
        
          <a href="../.." title="Home" class="md-flex md-footer-nav__link md-footer-nav__link--prev" rel="prev">
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-back md-footer-nav__button"></i>
            </div>
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Previous
                </span>
                Home
              </span>
            </div>
          </a>
        
        
          <a href="../snpdata/" title="SnpData" class="md-flex md-footer-nav__link md-footer-nav__link--next" rel="next">
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Next
                </span>
                SnpData
              </span>
            </div>
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-forward md-footer-nav__button"></i>
            </div>
          </a>
        
      </nav>
    </div>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        powered by
        <a href="https://www.mkdocs.org">MkDocs</a>
        and
        <a href="https://squidfunk.github.io/mkdocs-material/">
          Material for MkDocs</a>
      </div>
      
        
      
    </div>
  </div>
</footer>
      
    </div>
    
      <script src="../../assets/javascripts/application.a59e2a89.js"></script>
      
      <script>app.initialize({version:"0.17.5",url:{base:"../.."}})</script>
      
        <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
      
        <script src="../../assets/mathjaxhelper.js"></script>
      
    
    
      
    
  </body>
</html>