
<!DOCTYPE html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      <meta http-equiv="x-ua-compatible" content="ie=edge">
      
        <meta name="description" content="Compressed storage for SNP data">
      
      
      
        <meta name="author" content="Hua Zhou">
      
      
        <link rel="shortcut icon" href="../../assets/images/favicon.png">
      
      <meta name="generator" content="mkdocs-0.16.3, mkdocs-material-1.10.0">
    
    
      
        <title>HaplotypeArray - SnpArrays.jl</title>
      
    
    
      <script src="../../assets/javascripts/modernizr-e826f8942a.js"></script>
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/application-b6a6d08807.css">
      
        <link rel="stylesheet" href="../../assets/stylesheets/application-23f75ab9c7.palette.css">
      
    
    
      
        
        
        
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700|Roboto+Mono">
        <style>body,input{font-family:"Roboto","Helvetica Neue",Helvetica,Arial,sans-serif}code,kbd,pre{font-family:"Roboto Mono","Courier New",Courier,monospace}</style>
      
      <link rel="stylesheet" href="https://fonts.googleapis.com/icon?family=Material+Icons">
    
    
      <link rel="stylesheet" href="../../assets/Documenter.css">
    
    
  </head>
  
  
  
  
    <body data-md-color-primary="indigo" data-md-color-accent="blue">
  
    <svg class="md-svg">
      <defs>
        
        
          <svg xmlns="http://www.w3.org/2000/svg" width="416" height="448" viewBox="0 0 416 448" id="github"><path fill="currentColor" d="M160 304q0 10-3.125 20.5t-10.75 19T128 352t-18.125-8.5-10.75-19T96 304t3.125-20.5 10.75-19T128 256t18.125 8.5 10.75 19T160 304zm160 0q0 10-3.125 20.5t-10.75 19T288 352t-18.125-8.5-10.75-19T256 304t3.125-20.5 10.75-19T288 256t18.125 8.5 10.75 19T320 304zm40 0q0-30-17.25-51T296 232q-10.25 0-48.75 5.25Q229.5 240 208 240t-39.25-2.75Q130.75 232 120 232q-29.5 0-46.75 21T56 304q0 22 8 38.375t20.25 25.75 30.5 15 35 7.375 37.25 1.75h42q20.5 0 37.25-1.75t35-7.375 30.5-15 20.25-25.75T360 304zm56-44q0 51.75-15.25 82.75-9.5 19.25-26.375 33.25t-35.25 21.5-42.5 11.875-42.875 5.5T212 416q-19.5 0-35.5-.75t-36.875-3.125-38.125-7.5-34.25-12.875T37 371.5t-21.5-28.75Q0 312 0 260q0-59.25 34-99-6.75-20.5-6.75-42.5 0-29 12.75-54.5 27 0 47.5 9.875t47.25 30.875Q171.5 96 212 96q37 0 70 8 26.25-20.5 46.75-30.25T376 64q12.75 25.5 12.75 54.5 0 21.75-6.75 42 34 40 34 99.5z"/></svg>
        
      </defs>
    </svg>
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="drawer">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="search">
    <label class="md-overlay" data-md-component="overlay" for="drawer"></label>
    
      <header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid">
    <div class="md-flex">
      <div class="md-flex__cell md-flex__cell--shrink">
        <a href="../.." title="SnpArrays.jl" class="md-header-nav__button md-logo">
          
            <i class="md-icon md-icon--home"></i>
          
        </a>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        <label class="md-icon md-icon--menu md-header-nav__button" for="drawer"></label>
      </div>
      <div class="md-flex__cell md-flex__cell--stretch">
        <span class="md-flex__ellipsis md-header-nav__title">
          
            
              
                <span class="md-header-nav__parent">
                  Manual
                </span>
              
            
            HaplotypeArray
          
        </span>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        
          <label class="md-icon md-icon--search md-header-nav__button" for="search"></label>
          
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="search"></label>
  <div class="md-search__inner">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" required placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="query">
      <label class="md-icon md-search__icon" for="search"></label>
      <button type="reset" class="md-icon md-search__icon" data-md-component="reset">&#xE5CD;</button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="result" data-md-lang-search="" data-md-lang-tokenizer="[\s\-]+">
          <div class="md-search-result__meta" data-md-lang-result-none="No matching documents" data-md-lang-result-one="1 matching document" data-md-lang-result-other="# matching documents">
            Type to start searching
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
        
      </div>
      
        <div class="md-flex__cell md-flex__cell--shrink">
          <div class="md-header-nav__source">
            


  


  <a href="https://github.com/OpenMendel/SnpArrays.jl" title="Go to repository" class="md-source" data-md-source="github">
    
      <div class="md-source__icon">
        <svg viewBox="0 0 24 24" width="24" height="24">
          <use xlink:href="#github" width="24" height="24"></use>
        </svg>
      </div>
    
    <div class="md-source__repository">
      GitHub
    </div>
  </a>

          </div>
        </div>
      
    </div>
  </nav>
</header>
    
    <div class="md-container">
      
      
      <main class="md-main">
        <div class="md-main__inner md-grid" data-md-component="container">
          
            
              <div class="md-sidebar md-sidebar--primary" data-md-component="navigation">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    <nav class="md-nav md-nav--primary" data-md-level="0">
  <label class="md-nav__title md-nav__title--site" for="drawer">
    <div class="md-nav__button md-logo">
      
        <i class="md-icon md-icon--home"></i>
      
    </div>
    SnpArrays.jl
  </label>
  
    <div class="md-nav__source">
      


  


  <a href="https://github.com/OpenMendel/SnpArrays.jl" title="Go to repository" class="md-source" data-md-source="github">
    
      <div class="md-source__icon">
        <svg viewBox="0 0 24 24" width="24" height="24">
          <use xlink:href="#github" width="24" height="24"></use>
        </svg>
      </div>
    
    <div class="md-source__repository">
      GitHub
    </div>
  </a>

    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      


  <li class="md-nav__item">
    <a href="../.." title="Home" class="md-nav__link">
      Home
    </a>
  </li>

    
      
      
      

  


  <li class="md-nav__item md-nav__item--active md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-2" type="checkbox" id="nav-2" checked>
    
    <label class="md-nav__link" for="nav-2">
      Manual
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-2">
        Manual
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../snparray/" title="SnpArray" class="md-nav__link">
      SnpArray
    </a>
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../snpdata/" title="SnpData" class="md-nav__link">
      SnpData
    </a>
  </li>

        
          
          
          

  


  <li class="md-nav__item md-nav__item--active">
    
    <input class="md-toggle md-nav__toggle" data-md-toggle="toc" type="checkbox" id="toc">
    
      
    
    
      <label class="md-nav__link md-nav__link--active" for="toc">
        HaplotypeArray
      </label>
    
    <a href="./" title="HaplotypeArray" class="md-nav__link md-nav__link--active">
      HaplotypeArray
    </a>
    
      
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
    <label class="md-nav__title" for="toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#constructor" title="Constructor" class="md-nav__link">
    Constructor
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#summary-statistics" title="Summary statistics" class="md-nav__link">
    Summary statistics
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#subsetting-and-assignment" title="Subsetting and assignment" class="md-nav__link">
    Subsetting and assignment
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#copy-and-convert" title="Copy and convert" class="md-nav__link">
    Copy and convert
  </a>
  
</li>
      
      
      
    </ul>
  
</nav>
    
  </li>

        
      </ul>
    </nav>
  </li>

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              <div class="md-sidebar md-sidebar--secondary" data-md-component="toc">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
    <label class="md-nav__title" for="toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#constructor" title="Constructor" class="md-nav__link">
    Constructor
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#summary-statistics" title="Summary statistics" class="md-nav__link">
    Summary statistics
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#subsetting-and-assignment" title="Subsetting and assignment" class="md-nav__link">
    Subsetting and assignment
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#copy-and-convert" title="Copy and convert" class="md-nav__link">
    Copy and convert
  </a>
  
</li>
      
      
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/OpenMendel/SnpArrays.jl/edit/master/docs/man/haplotypearray.md" title="Edit this page" class="md-icon md-content__icon">&#xE3C9;</a>
                
                
                <p><a id='HaplotypeArray-1'></a></p>
<h1 id="haplotypearray">HaplotypeArray</h1>
<p>The <code>HaplotypeArray</code> type is similar to <a href="../snparray/#SnpArray-1">SnpArray</a>, but with two key differences:  </p>
<ul>
<li>The two alleles are ordered, therefore (A1, A2) is different from (A2, A1);</li>
<li>The code <code>(true,false)</code> means the genotype (A2, A1), instead of the missing genotype in <code>SnpArray</code>.</li>
</ul>
<table>
<thead>
<tr>
<th align="center">Genotype</th>
<th align="center">HaplotypeArray</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center">A1,A1</td>
<td align="center">(false,false)</td>
</tr>
<tr>
<td align="center">A1,A2</td>
<td align="center">(false,true)</td>
</tr>
<tr>
<td align="center">A2,A1</td>
<td align="center">(true,false)</td>
</tr>
<tr>
<td align="center">A2,A2</td>
<td align="center">(true,true)</td>
</tr>
</tbody>
</table>
<p>Each bit <code>true</code> in <code>HaplotypeArray</code> indicates a copy of the A2 allele. </p>
<p><a id='Constructor-1'></a></p>
<h2 id="constructor">Constructor</h2>
<p>There are various ways to initialize a <code>HaplotypeArray</code>.  </p>
<ul>
<li><code>HaplotypeArray</code> can be intialized from two <code>BitArray</code>s. Each <code>BitArray</code> indicates an A2 allele copy in the first and second positions respectively.</li>
</ul>
<div class="codehilite"><pre><span></span><span class="k">using</span> <span class="n">SnpArrays</span>
<span class="n">h</span> <span class="o">=</span> <span class="n">HaplotypeArray</span><span class="p">(</span><span class="n">bitrand</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span> <span class="n">bitrand</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>
</pre></div>


<div class="codehilite"><pre><span></span>5×3 SnpArrays.HaplotypeArray{2}:
 (true, false)  (false, true)   (true, false) 
 (true, true)   (false, false)  (true, false) 
 (true, false)  (true, true)    (true, false) 
 (true, false)  (false, false)  (false, false)
 (false, true)  (true, false)   (false, false)
</pre></div>


<ul>
<li><code>HaplotypeArray</code> can be intialized from a <code>SnpArray</code>.</li>
</ul>
<div class="codehilite"><pre><span></span><span class="n">s</span> <span class="o">=</span> <span class="n">SnpArray</span><span class="p">(</span><span class="n">bitrand</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">3</span><span class="p">),</span> <span class="n">bitrand</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">3</span><span class="p">))</span>
<span class="n">h</span> <span class="o">=</span> <span class="n">HaplotypeArray</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
</pre></div>


<div class="codehilite"><pre><span></span>5×3 SnpArrays.HaplotypeArray{2}:
 (false, true)   (true, true)    (false, true) 
 (true, false)   (true, false)   (true, true)  
 (false, false)  (false, false)  (false, true) 
 (false, false)  (false, false)  (true, false) 
 (true, true)    (true, false)   (false, false)
</pre></div>


<p>This constructor does <strong>not</strong> copy data from <code>SnpArray</code>. Therefore both <code>h</code> and <code>s</code> points to the same patch of memory. Only the interpretation of <code>(true, false)</code> changes.</p>
<div class="codehilite"><pre><span></span><span class="n">isnan</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
</pre></div>


<div class="codehilite"><pre><span></span>5×3 BitArray{2}:
 false  false  false
  true   true  false
 false  false  false
 false  false   true
 false   true  false
</pre></div>


<div class="codehilite"><pre><span></span><span class="n">isnan</span><span class="p">(</span><span class="n">h</span><span class="p">)</span>
</pre></div>


<div class="codehilite"><pre><span></span>5×3 BitArray{2}:
 false  false  false
 false  false  false
 false  false  false
 false  false  false
 false  false  false
</pre></div>


<p>Changes to <code>h</code> also effect <code>s</code>.</p>
<div class="codehilite"><pre><span></span><span class="n">h</span><span class="p">[</span><span class="n">isnan</span><span class="p">(</span><span class="n">s</span><span class="p">)]</span> <span class="o">=</span> <span class="p">(</span><span class="kc">true</span><span class="p">,</span> <span class="kc">true</span><span class="p">)</span>
<span class="n">isnan</span><span class="p">(</span><span class="n">s</span><span class="p">)</span>
</pre></div>


<div class="codehilite"><pre><span></span>5×3 BitArray{2}:
 false  false  false
 false  false  false
 false  false  false
 false  false  false
 false  false  false
</pre></div>


<ul>
<li><code>HaplotypeArray(m, n)</code> generates an m by n <code>HaplotypeArray</code> of all A1 alleles.</li>
</ul>
<div class="codehilite"><pre><span></span><span class="n">HaplotypeArray</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>
</pre></div>


<div class="codehilite"><pre><span></span>5×3 SnpArrays.HaplotypeArray{2}:
 (false, false)  (false, false)  (false, false)
 (false, false)  (false, false)  (false, false)
 (false, false)  (false, false)  (false, false)
 (false, false)  (false, false)  (false, false)
 (false, false)  (false, false)  (false, false)
</pre></div>


<p><a id='Summary-statistics-1'></a></p>
<h2 id="summary-statistics">Summary statistics</h2>
<p><code>summarize</code> when applied to a <code>HaplotypeArray</code> only computes  </p>
<ul>
<li><code>maf</code>: minor allele frequencies, taking into account of missingness.</li>
<li><code>minor_allele</code>: a <code>BitVector</code> indicating the minor allele for each SNP.   <code>minor_allele[j]==true</code> means A1 is the minor allele for SNP j; <code>minor_allele[j]==false</code> means A2 is the minor allele for SNP j.</li>
</ul>
<div class="codehilite"><pre><span></span><span class="n">maf</span><span class="p">,</span> <span class="n">minor_allele</span> <span class="o">=</span> <span class="n">summarize</span><span class="p">(</span><span class="n">h</span><span class="p">)</span>
</pre></div>


<div class="codehilite"><pre><span></span>([0.5, 0.4, 0.4], Bool[false, true, true])
</pre></div>


<p><a id='Subsetting-and-assignment-1'></a></p>
<h2 id="subsetting-and-assignment">Subsetting and assignment</h2>
<p>Subsetting and assignment work the same as <a href="../snparray/#SnpArray-1">SnpArray</a>.</p>
<p><a id='Copy-and-convert-1'></a></p>
<h2 id="copy-and-convert">Copy and convert</h2>
<p>Copying or converting a <code>HaplotypeArray</code> or slices of it to numeric arrays of <strong>minor allele counts</strong> is similar to <code>SnpArray</code> with the exception there are no missing genotypes in <code>HaplotypeArray</code>. So the keyword <code>impute</code> is not relevant anymore.</p>
<div class="codehilite"><pre><span></span><span class="c"># convert to Matrix{Float64}</span>
<span class="n">h_f64</span> <span class="o">=</span> <span class="n">convert</span><span class="p">(</span><span class="kt">Matrix</span><span class="p">{</span><span class="kt">Float64</span><span class="p">},</span> <span class="n">h</span><span class="p">)</span>
</pre></div>


<div class="codehilite"><pre><span></span>5×3 Array{Float64,2}:
 1.0  0.0  1.0
 2.0  0.0  0.0
 0.0  2.0  1.0
 0.0  2.0  0.0
 2.0  0.0  2.0
</pre></div>


<p>By default <code>convert</code> translates genotypes according to the <em>additive</em> SNP model, which essentially counts the number of <strong>minor allele</strong> (0, 1 or 2) per genotype. Other SNP models are <em>dominant</em> and <em>recessive</em>, both in terms of the <strong>minor allele</strong>. When <code>A1</code> is the minor allele, genotypes are translated to real number according to</p>
<table>
<thead>
<tr>
<th align="center">Genotype</th>
<th align="center"><code>HaplotypeArray</code></th>
<th align="center"><code>model=:additive</code></th>
<th align="center"><code>model=:dominant</code></th>
<th align="center"><code>model=:recessive</code></th>
</tr>
</thead>
<tbody>
<tr>
<td align="center">A1,A1</td>
<td align="center">00</td>
<td align="center">2</td>
<td align="center">1</td>
<td align="center">1</td>
</tr>
<tr>
<td align="center">A1,A2</td>
<td align="center">01</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">0</td>
</tr>
<tr>
<td align="center">A2,A1</td>
<td align="center">10</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">0</td>
</tr>
<tr>
<td align="center">A2,A2</td>
<td align="center">11</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
</tbody>
</table>
<p>When <code>A2</code> is the minor allele, genotypes are translated according to</p>
<table>
<thead>
<tr>
<th align="center">Genotype</th>
<th align="center"><code>HaplotypeArray</code></th>
<th align="center"><code>model=:additive</code></th>
<th align="center"><code>model=:dominant</code></th>
<th align="center"><code>model=:recessive</code></th>
</tr>
</thead>
<tbody>
<tr>
<td align="center">A1,A1</td>
<td align="center">00</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">0</td>
</tr>
<tr>
<td align="center">A1,A2</td>
<td align="center">01</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">0</td>
</tr>
<tr>
<td align="center">A2,A1</td>
<td align="center">01</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">0</td>
</tr>
<tr>
<td align="center">A2,A2</td>
<td align="center">11</td>
<td align="center">2</td>
<td align="center">1</td>
<td align="center">1</td>
</tr>
</tbody>
</table>
<div class="codehilite"><pre><span></span><span class="p">[</span><span class="n">convert</span><span class="p">(</span><span class="kt">Vector</span><span class="p">{</span><span class="kt">Float64</span><span class="p">},</span> <span class="n">h</span><span class="p">[</span><span class="mi">1</span><span class="o">:</span><span class="mi">5</span><span class="p">,</span> <span class="mi">3</span><span class="p">];</span> <span class="n">model</span> <span class="o">=</span> <span class="o">:</span><span class="n">additive</span><span class="p">)</span> <span class="n">convert</span><span class="p">(</span><span class="kt">Vector</span><span class="p">{</span><span class="kt">Float64</span><span class="p">},</span> <span class="n">h</span><span class="p">[</span><span class="mi">1</span><span class="o">:</span><span class="mi">5</span><span class="p">,</span> <span class="mi">3</span><span class="p">];</span> <span class="n">model</span> <span class="o">=</span> <span class="o">:</span><span class="n">dominant</span><span class="p">)</span> <span class="n">convert</span><span class="p">(</span><span class="kt">Vector</span><span class="p">{</span><span class="kt">Float64</span><span class="p">},</span> <span class="n">h</span><span class="p">[</span><span class="mi">1</span><span class="o">:</span><span class="mi">5</span><span class="p">,</span> <span class="mi">3</span><span class="p">];</span> <span class="n">model</span> <span class="o">=</span> <span class="o">:</span><span class="n">recessive</span><span class="p">)]</span>
</pre></div>


<div class="codehilite"><pre><span></span>5×3 Array{Float64,2}:
 1.0  1.0  0.0
 0.0  0.0  0.0
 1.0  1.0  0.0
 0.0  0.0  0.0
 2.0  1.0  1.0
</pre></div>


<p>By default <code>convert</code> does <strong>not</strong> center and scale genotypes. Setting the optional arguments <code>center=true, scale=true</code> centers genotypes at 2MAF and scales them by $[2 \cdot \text{MAF} \cdot (1 - \text{MAF})]^{-1/2}$. Mono-allelic SNPs (MAF=0) are not scaled.</p>
<div class="codehilite"><pre><span></span><span class="p">[</span><span class="n">convert</span><span class="p">(</span><span class="kt">Vector</span><span class="p">{</span><span class="kt">Float64</span><span class="p">},</span> <span class="n">h</span><span class="p">[</span><span class="o">:</span><span class="p">,</span> <span class="mi">3</span><span class="p">])</span> <span class="n">convert</span><span class="p">(</span><span class="kt">Vector</span><span class="p">{</span><span class="kt">Float64</span><span class="p">},</span> <span class="n">h</span><span class="p">[</span><span class="o">:</span><span class="p">,</span> <span class="mi">3</span><span class="p">];</span> <span class="n">center</span> <span class="o">=</span> <span class="kc">true</span><span class="p">,</span> <span class="n">scale</span> <span class="o">=</span> <span class="kc">true</span><span class="p">)]</span>
</pre></div>


<div class="codehilite"><pre><span></span>5×2 Array{Float64,2}:
 1.0   0.288675
 0.0  -1.1547  
 1.0   0.288675
 0.0  -1.1547  
 2.0   1.73205
</pre></div>


<p><code>copy!</code> is the in-place version of <code>convert()</code>. Options such as GWAS loop over SNPs and perform statistical anlaysis for each SNP. This can be achieved by</p>
<div class="codehilite"><pre><span></span><span class="n">g</span> <span class="o">=</span> <span class="n">zeros</span><span class="p">(</span><span class="n">size</span><span class="p">(</span><span class="n">h</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
<span class="k">for</span> <span class="n">j</span> <span class="o">=</span> <span class="mi">1</span><span class="o">:</span><span class="n">size</span><span class="p">(</span><span class="n">h</span><span class="p">,</span> <span class="mi">2</span><span class="p">)</span>
    <span class="n">copy!</span><span class="p">(</span><span class="n">g</span><span class="p">,</span> <span class="n">h</span><span class="p">[</span><span class="o">:</span><span class="p">,</span> <span class="n">j</span><span class="p">];</span> <span class="n">model</span> <span class="o">=</span> <span class="o">:</span><span class="n">additive</span><span class="p">)</span>
    <span class="c"># do statistical anlaysis</span>
<span class="k">end</span>
</pre></div>
                
                  
                
              
              
                
              
            </article>
          </div>
        </div>
      </main>
      
        
<footer class="md-footer">
  
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid">
        
          <a href="../snpdata/" title="SnpData" class="md-flex md-footer-nav__link md-footer-nav__link--prev" rel="prev">
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-back md-footer-nav__button"></i>
            </div>
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Previous
                </span>
                SnpData
              </span>
            </div>
          </a>
        
        
      </nav>
    </div>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        powered by
        <a href="http://www.mkdocs.org" title="MkDocs">MkDocs</a>
        and
        <a href="http://squidfunk.github.io/mkdocs-material/" title="Material for MkDocs">
          Material for MkDocs</a>
      </div>
      
        
      
    </div>
  </div>
</footer>
      
    </div>
    
      <script src="../../assets/javascripts/application-f3ab9e5ff8.js"></script>
      
      
      <script>app.initialize({url:{base:"../.."}})</script>
      
        <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS_HTML"></script>
      
        <script src="../../assets/mathjaxhelper.js"></script>
      
    
    
      
    
  </body>
</html>